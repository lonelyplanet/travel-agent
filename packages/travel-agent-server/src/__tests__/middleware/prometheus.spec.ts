const promBundleMock = jest.fn();
const setDefaultLabelsMock = jest.fn();
const collectDefaultMetricsMock = jest.fn();
jest.mock("express-prom-bundle", () => promBundleMock);

import * as promBundle from "express-prom-bundle";
promBundle.promClient = {
  register: {
    setDefaultLabels: setDefaultLabelsMock,
  },
  collectDefaultMetrics: collectDefaultMetricsMock,
};

import createPrometheusMiddleware, { customNormalize } from "../../middleware/prometheus";

describe("prometheusMiddleware", () => {
  describe("customNormalize", () => {
    it("should return the name of the first matching route", () => {
      const url = "usa/nashville/restaurants/bastion/a/poi-eat/1534928/362228";
      const routes = [
        { route: "(.*)/news/:date/:slug", name: "News#show" },
        { route: "(.*)/a/:type-:kind/:id/:placeId", name: "Poi#show" },
      ];

      const path = customNormalize(url, routes);

      expect(path).toBe("Poi#show");
    });
  });

  it("should return the provided default path if the url doesn't match a route", () => {
    const defaultPath = "news/";
    const url = "usa/nashville/restaurants/bastion/a/poi-eat/1534928/362228";
    const routes = [
      { route: "(.*)/news/:date/:slug", name: "News#show" },
    ];

    const path = customNormalize(url, routes, { defaultPath });

    expect(path).toBe(defaultPath);
  });

  it("should return the provided url if the url doesn't match a route", () => {
    const defaultPath = "news/";
    const url = "usa/nashville/restaurants/bastion/a/poi-eat/1534928/362228";
    const routes = [
      { route: "(.*)/news/:date/:slug", name: "News#show" },
    ];

    const path = customNormalize(url, routes, {});

    expect(path).toBe(url);
  });

  describe("createPrometheusMiddleware", () => {
    it("should return the name of the first matching route", () => {
      const url = "usa/nashville/restaurants/bastion/a/poi-eat/1534928/362228";
      const routes = [
        { route: "(.*)/a/:type-:kind/:id/:placeId", name: "Poi#show" },
      ];

      const bundle = createPrometheusMiddleware({ routes, defaultPath: "test/" });

      expect(setDefaultLabelsMock).toHaveBeenCalledWith({ direction: "inbound" });
      expect(collectDefaultMetricsMock).toHaveBeenCalledTimes(1);
      expect(promBundle.normalizePath({ url })).toBe("Poi#show");
      expect(promBundleMock.mock.calls[0]).toMatchSnapshot();
    });
  });
});
