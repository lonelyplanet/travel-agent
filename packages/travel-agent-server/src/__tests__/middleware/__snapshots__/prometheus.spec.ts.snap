// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prometheusMiddleware createPrometheusMiddleware should return the name of the first matching route 1`] = `
Array [
  Object {
    "buckets": Array [
      0.01,
      0.05,
      0.1,
      0.2,
      0.3,
      0.4,
      0.5,
      0.6,
      0.7,
      0.8,
      0.9,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ],
    "excludeRoutes": Array [
      /assets/,
    ],
    "includeMethod": true,
    "includePath": true,
  },
]
`;
